File: express-service/src/routes/zipcode.ts
import { Router, Request, Response } from 'express';
import { AddressValidator } from '../services/addressValidator';
import { ZipcodeCityRequest } from '../models/validation';

/**
 * Regex for valid zipcode format:
 * - Exactly 5 digits (e.g., "20500")
 * - Or 5 digits + hyphen + 4 digits (e.g., "20500-0001")
 */
const ZIPCODE_REGEX = /^\d{5}(-\d{4})?$/;

/**
 * Creates the zipcode-city verification router.
 * POST /validate/zipcode-city
 */
export function createZipcodeRouter(addressValidator: AddressValidator): Router {
  const router = Router();

  router.post('/validate/zipcode-city', async (req: Request, res: Response) => {
    const { zipcode, city } = req.body as ZipcodeCityRequest;

    // Validate zipcode format
    if (!zipcode || !ZIPCODE_REGEX.test(zipcode)) {
      return res.status(400).json({
        detail: {
          error_code: 'INVALID_ZIPCODE_FORMAT',
          message: 'Zipcode must be in 5-digit format (e.g., "20500") or 5+4 format (e.g., "20500-0001")',
        },
      });
    }

    // Validate city is provided
    if (!city) {
      return res.status(400).json({
        detail: {
          error_code: 'VALIDATION_ERROR',
          message: 'City is required',
        },
      });
    }

    try {
      const result = await addressValidator.verifyZipcodeCity(zipcode, city);
      return res.status(200).json(result);
    } catch (error) {
      return res.status(500).json({
        detail: {
          error_code: 'INTERNAL_ERROR',
          message: 'An unexpected error occurred during zipcode-city verification',
        },
      });
    }
  });

  return router;
}

File: python-service/app/routes/zipcode.py
"""Route for zipcode-city verification endpoint.

Exposes POST /validate/zipcode-city which accepts a zipcode and city,
validates the zipcode format, and delegates to AddressValidator.
"""

import re

from fastapi import APIRouter, Depends, HTTPException

from app.adapters.usps_adapter import USPSAdapterInterface, get_usps_adapter
from app.models.validation import ZipcodeCityRequest, ZipcodeCityResult
from app.services.address_validator import AddressValidator

router = APIRouter()

# Regex pattern for valid zipcode: exactly 5 digits, or 5 digits + hyphen + 4 digits
_ZIPCODE_PATTERN = re.compile(r"^\d{5}(-\d{4})?$")


def _get_address_validator(
    adapter: USPSAdapterInterface = Depends(get_usps_adapter),
) -> AddressValidator:
    """Dependency that provides an AddressValidator wired with the USPS adapter."""
    return AddressValidator(usps_adapter=adapter)


@router.post("/validate/zipcode-city", response_model=ZipcodeCityResult)
async def verify_zipcode_city(
    request: ZipcodeCityRequest,
    validator: AddressValidator = Depends(_get_address_validator),
) -> ZipcodeCityResult:
    """Verify that a zipcode matches a given city via USPS.

    Accepts a ZipcodeCityRequest (zipcode, city) in the request body.
    Validates that the zipcode is in 5-digit or 5+4 format. If invalid,
    returns a 400 error with the INVALID_ZIPCODE_FORMAT error code.

    On success, delegates to AddressValidator.verify_zipcode_city() and
    returns the ZipcodeCityResult.
    """
    if not _ZIPCODE_PATTERN.match(request.zipcode):
        raise HTTPException(
            status_code=400,
            detail={
                "error_code": "INVALID_ZIPCODE_FORMAT",
                "message": "Zipcode must be 5 digits or 5+4 format",
            },
        )

    return await validator.verify_zipcode_city(request.zipcode, request.city)
